### PR TITLE
Cart feature

### PR DESCRIPTION
**Pull request by Iamyuvraj143**
Add to Cart Feature

<kpd>cart</kpd>

<kpd>add to cart form</kpd>

### CODE REVIEW

File: app/controllers/cart_products_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+class CartProductsController < ApplicationController
+  before_action :auth_check
```
Comment (kalpana19sharma) [inline app/controllers/cart_products_controller.rb:2]: Are you using this check in other controllers as well? If yes then move the method to application controller.

File: app/controllers/cart_products_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  before_action :load_cart, only: %i( new create edit update destroy)
+
+  def new
+    @cart_product = @cart.cart_products.new
+  end
+
+  def create
+    @cart_product = @cart.cart_products.new(cart_params)
+    @cart_product.cart_product_total = @cart_product.product.price * @cart_product.quantity
```
Comment (kalpana19sharma) [inline app/controllers/cart_products_controller.rb]: Don't use table name prefix for attributes. Its should be just `@cart_product.total`

File: app/controllers/cart_products_controller.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      redirect_to shopping_cart_index_path status: 303
+      flash[:notice] = "added to cart"
+    else
+      render :new 
+    end
+  end
+
+
+  def destroy
+    @cart_product = @cart.cart_products.find_by(id: params[:id])
```
Comment (kalpana19sharma) [inline app/controllers/cart_products_controller.rb:22]: Handle the case if cart product not found

File: app/controllers/registrations_controller.rb [hunk 4]
```diff
@@ -6,7 +6,9 @@ def new
   
   def create
     @user = User.new(user_params)
+    @cart = ShoppingCart.new
       if @user.save
+        @user.shopping_cart = @cart
```
Comment (kalpana19sharma) [inline app/controllers/registrations_controller.rb]: Use a callback for creating cart. What if I create a user from rails console?

File: app/controllers/shopping_cart_controller.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class ShoppingCartController < ApplicationController
+  before_action :auth_check
+  
+  def index
+    user = Current.user
+    cart = user.shopping_cart
+    @cart_items = cart.cart_products.all
```
Comment (kalpana19sharma) [inline app/controllers/shopping_cart_controller.rb]:
```suggestion
    @cart_products = cart.cart_products.all
```

File: app/controllers/shopping_cart_controller.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class ShoppingCartController < ApplicationController
+  before_action :auth_check
+  
+  def index
+    user = Current.user
+    cart = user.shopping_cart
```
Comment (kalpana19sharma) [inline app/controllers/shopping_cart_controller.rb]:
```suggestion
    cart = Current.user.shopping_cart
```

File: app/helpers/shopping_cart_helper.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+module ShoppingCartHelper
+end
```
Comment (kalpana19sharma) [inline app/helpers/shopping_cart_helper.rb:2]: Remove this file

File: app/models/user.rb [hunk 8]
```diff
@@ -1,5 +1,6 @@
 class User < ApplicationRecord
   # adds virtual attributes for authentication
+  has_one :shopping_cart
```
Comment (kalpana19sharma) [inline app/models/user.rb]: Should we use dependent destroy here?

File: app/models/shopping_cart.rb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+class ShoppingCart < ApplicationRecord
+  has_many :cart_products
```
Comment (kalpana19sharma) [inline app/models/shopping_cart.rb]: Should we use dependent destroy here?

File: config/routes.rb [hunk 10]
```diff
@@ -1,6 +1,8 @@
 Rails.application.routes.draw do
   root 'welcome#index'
   resources :products
+  resources :shopping_cart, only: [:index]
+  resources :cart_products
```
Comment (kalpana19sharma) [inline config/routes.rb]: Are we using all of the actions?

File: db/migrate/20220422101301_create_cart_products.rb [hunk 11]
```diff
@@ -0,0 +1,11 @@
+class CreateCartProducts < ActiveRecord::Migration[6.1]
+  def change
+    create_table :cart_products do |t|
+      t.integer :quantity
+      t.float :cart_product_total
```
Comment (kalpana19sharma) [inline db/migrate/20220422101301_create_cart_products.rb:5]:
```suggestion
      t.float :total
```

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

File: app/controllers/registrations_controller.rb (hunk 4 above)
Comment (Iamyuvraj143) [reply app/controllers/registrations_controller.rb]: Done

File: app/controllers/shopping_cart_controller.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
```
Comment (kalpana19sharma) [inline app/controllers/shopping_cart_controller.rb]: What's the need of `user` local variable here?

**Review by kalpana19sharma — CHANGES_REQUESTED** [review]

**Review by kalpana19sharma — APPROVED** [review]

**Pull request merged**